src/extract/faster_bottom_up.rs

```diff
@@ -0,0 +1,117 @@
+use super::*;
+
+/// A faster bottom up extractor inspired by the faster-greedy-dag extractor.
+pub struct BottomUpExtractor;
+
+impl Extractor for BottomUpExtractor {
+    fn extract(&self, egraph: &EGraph, _roots: &[ClassId]) -> ExtractionResult {
+        // 1. build map from class to parent nodes
+        let mut parents = IndexMap::<ClassId, Vec<NodeId>>::default();
+        let n2c = |nid: &NodeId| egraph.nid_to_cid(nid);
+
+        for class in egraph.classes().values() {
+            parents.insert(class.id.clone(), Vec::new());
+        }
+        for class in egraph.classes().values() {
+            for node in &class.nodes {
+                for c in &egraph[node].children {
+                    parents[n2c(c)].push(node.clone());
+                }
+            }
+        }
+
+        // 2. start analysis from leaves
+        let mut analysis_pending = UniqueQueue::default();
+
+        for class in egraph.classes().values() {
+            for node in &class.nodes {
+                if egraph[node].is_leaf() {
+                    analysis_pending.insert(node.clone());
+                }
+            }
+        }
+
+        let mut result = ExtractionResult::default();
+        let mut costs = IndexMap::<ClassId, Cost>::default();
+
+        while let Some(node_id) = analysis_pending.pop() {
+            let class_id = n2c(&node_id);
+            let node = &egraph[&node_id];
+            if node.children.iter().all(|c| costs.contains_key(n2c(c))) {
+                let prev_cost = costs.get(class_id).unwrap_or(&INFINITY);
+
+                let cost = result.node_sum_cost(egraph, node, &costs);
+                if cost < *prev_cost {
+                    result.choose(class_id.clone(), node_id.clone());
+                    costs.insert(class_id.clone(), cost);
+                    analysis_pending.extend(parents[class_id].iter().cloned());
+                }
+            } else {
```

Comment:
Why do we need this? 

Right now, there's at least one child that doesn't currently have a cost. That means that at some point if we can extract this node at all we number of "unextracted children" will go from 1 to 0. At that point, this node would get re-added, no?

Reply:
Good point! That actually made it significantly faster!
It also made `faster-greedy-dag` faster
```
cumulative time for faster-bottom-up: 1155ms
cumulative time for faster-greedy-dag: 1834ms
```